File: IdentityX/testing_framework/refresh_test.go
package testing

import (
	"GoAuth/internal/apierr"
	"net/http"
	"sync"
	"testing"
	"time"

	"github.com/stretchr/testify/require"
)

func testRefresh(t *testing.T, suite *TestSuite) {
	t.Run("RefreshSuccess", func(t *testing.T) {
		client := suite.NewClient(t)
		user := client.WithCredentials("refresh_success@mail.com", ValidPassword).Register().Login()

		oldAccess := user.auth.AccessToken
		oldRefresh := user.auth.RefreshToken

		if user.auth == nil {
			t.Errorf("Refresh called on unauthenticated client")
		}

		resp := user.POST("/auth/refresh").
			WithCookie("refresh_token", user.auth.RefreshToken).
			Expect(http.StatusOK)

		access := resp.resp.Cookie("access_token")
		refresh := resp.resp.Cookie("refresh_token")

		if access.Raw() == nil {
			t.Errorf("Expected access cookie after refresh but got nil")
		}

		if refresh.Raw() == nil {
			t.Errorf("Expected refresh cookie after refresh but got nil")
		}

		refreshed := user.WithAuth(&AuthContext{
			AccessToken:  access.Value().Raw(),
			RefreshToken: refresh.Value().Raw(),
		})

		require.NotEqual(t, oldAccess, refreshed.auth.AccessToken, "Access token should change after refresh")
		require.NotEqual(t, oldRefresh, refreshed.auth.RefreshToken, "Refresh token should change after refresh")
	})

	t.Run("UseOldTokenError", func(t *testing.T) {
		client := suite.NewClient(t)
		oldAuth := client.WithCredentials("old_token_user@mail.com", ValidPassword).Register().Login()

		oldAuth.Refresh() // Refresh the token, invalidating the old one, never save the new authed user

		// Now try to use the old access token
		client.WithAuth(oldAuth.auth).GET("/sessions").
			Expect(http.StatusUnauthorized).
			HasErrID(apierr.TokenReuseIdentified).
			HasMessage("refresh token reuse not allowed")
	})

	t.Run("RefreshRevokedToken", func(t *testing.T) {
		client := suite.NewClient(t)
		user := client.WithCredentials("revoked_refresh_user@mail.com", ValidPassword).Register().Login()
		refreshToken := user.auth.RefreshToken

		// Manually revoke the session in the database
		_, err := suite.DB.Exec(`
			UPDATE sessions 
			SET revoked_at = NOW() 
			WHERE token_id = (
				SELECT token_id 
				FROM sessions 
				WHERE identity_id = (
					SELECT id 
					FROM identities 
					WHERE entity_id = (
						SELECT id 
						FROM users 
						WHERE email = 'revoked_refresh_user@mail.com'
					)
				) LIMIT 1
			)
		`)
		require.NoError(t, err)

		// Attempt to use the revoked refresh token
		deniedClient := suite.NewClient(t)
		deniedClient.POST("/auth/refresh").
			WithCookie("refresh_token", refreshToken).
			Expect(http.StatusUnauthorized).
			HasErrID(apierr.SessionNotFound).
			HasMessage("session not found or revoked")
	})

	t.Run("ConcurrentRefresh", func(t *testing.T) {
		concUser := suite.NewClient(t).WithCredentials("concurrent_refresh@mail.com", ValidPassword).Register().Login()
		refreshToken := concUser.auth.RefreshToken

		concurrency := 5
		results := make(chan int, concurrency)
		httpClient := &http.Client{Timeout: 10 * time.Second}
		var wg sync.WaitGroup
		for i := 0; i < concurrency; i++ {
			wg.Add(1)
			go func() {
				defer wg.Done()
				req, _ := http.NewRequest("POST", suite.Server.URL+"/auth/refresh", nil)
				req.AddCookie(&http.Cookie{Name: "refresh_token", Value: refreshToken})
				resp, err := httpClient.Do(req)
				if err != nil {
					results <- 0
					return
				}
				defer resp.Body.Close()
				results <- resp.StatusCode
			}()
		}

		wg.Wait()
		close(results)

		successCount := 0
		failCount := 0
		for status := range results {
			if status == http.StatusOK {
				successCount++
			} else if status == http.StatusUnauthorized {
				failCount++
			}
		}

		require.Equal(t, 1, successCount, "Only one refresh request should succeed")
		require.Equal(t, concurrency-1, failCount, "All other refresh requests should fail due to token reuse")
	})
}

File: IdentityX/testing_framework/refresh.test.ts
import { afterAll, describe, expect, test } from "vitest";
import { createClient } from "./helpers/index.js";
import { assertErrID, assertMessage, shouldFail } from "./helpers/assert.js";
import { dbQuery, closeDB } from "./helpers/db.js";
import { ValidPassword } from "./fixtures/auth/testdata.js";

// Error IDs
// TokenReuseIdentified = fail.ID(0, "TOKEN", 18, false, ...) → 0_TOKEN_0018_D
// SessionNotFound      = fail.ID(0, "SESSION", 0, true, ...)  → 0_SESSION_0000_S
const ErrTokenReuseIdentified = "0_TOKEN_0018_D";
const ErrSessionNotFound      = "0_SESSION_0000_S";

afterAll(async () => {
    await closeDB();
});

// ============================================================================
// REFRESH TESTS
// ============================================================================

describe("Refresh", () => {
    test("RefreshSuccess", async () => {
        await createClient().withCredentials("refresh_success@mail.com", ValidPassword).register();
        const user = await createClient().withCredentials("refresh_success@mail.com", ValidPassword).login();

        const oldAccess  = user.auth!.accessToken;
        const oldRefresh = user.auth!.refreshToken;

        const refreshed = await user.refresh();

        expect(refreshed.auth!.accessToken).not.toBe(oldAccess);
        expect(refreshed.auth!.refreshToken).not.toBe(oldRefresh);
    });

    test("UseOldTokenError", async () => {
        await createClient().withCredentials("old_token_user@mail.com", ValidPassword).register();
        const oldAuth = await createClient().withCredentials("old_token_user@mail.com", ValidPassword).login();

        // Refresh but discard the new auth — old tokens are now invalidated
        await oldAuth.refresh();

        // Try to use the old access token
        const body = await shouldFail(oldAuth.get("/sessions"), 401);
        assertErrID(body, ErrTokenReuseIdentified);
        assertMessage(body, "refresh token reuse not allowed");
    });

    test("RefreshRevokedToken", async () => {
        await createClient().withCredentials("revoked_refresh_user@mail.com", ValidPassword).register();
        const user = await createClient().withCredentials("revoked_refresh_user@mail.com", ValidPassword).login();
        const refreshToken = user.auth!.refreshToken;

        // Manually revoke the session in the database
        await dbQuery(`
      UPDATE sessions
      SET revoked_at = NOW()
      WHERE token_id = (
        SELECT token_id FROM sessions
        WHERE identity_id = (
          SELECT id FROM identities
          WHERE entity_id = (
            SELECT id FROM users WHERE email = 'revoked_refresh_user@mail.com'
          )
        ) LIMIT 1
      )
    `);

        // Attempt to use the revoked refresh token
        const denied = createClient();
        const body = await shouldFail(
            denied.http.post("/auth/refresh", {}, {
                headers: { Cookie: `refresh_token=${refreshToken}` },
            }),
            401
        );
        assertErrID(body, ErrSessionNotFound);
        assertMessage(body, "session not found or revoked");
    });

    test("ConcurrentRefresh", async () => {
        await createClient().withCredentials("concurrent_refresh@mail.com", ValidPassword).register();
        const user = await createClient().withCredentials("concurrent_refresh@mail.com", ValidPassword).login();
        const refreshToken = user.auth!.refreshToken;

        const concurrency = 5;

        const results = await Promise.all(
            Array.from({ length: concurrency }, () =>
                createClient().http.post("/auth/refresh", {}, {
                    headers: { Cookie: `refresh_token=${refreshToken}` },
                    validateStatus: () => true,
                }).then(res => res.status)
            )
        );

        const successCount = results.filter(s => s === 200).length;
        const failCount    = results.filter(s => s === 401).length;

        expect(successCount).toBe(1);
        expect(failCount).toBe(concurrency - 1);
    });
});